### PR TITLE
Handle cancellation in `DatabasePool.runInteraction()`

### PR DESCRIPTION
**Pull request by squahtx**
To handle cancellation, we ensure that `after_callback`s and
`exception_callback`s are always run, since the transaction will
complete on another thread regardless of cancellation.

We also wait until everything is done before releasing the
`CancelledError`, so that logging contexts won't get used after they
have been finished.

Part of #3528.
Depends on and contains changes from two other PRs: #12180 and #12198. Please ignore the first two commits.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by clokep** [comment]
Pretty certain this looks reasonable, but depends on the other PRs! 👍

**Comment by erikjohnston** [comment]
Am I write in thinking that for function wrapped in e.g. `@cached` the cancellation won't propagate to the inner `runInteraction`?

**Comment by squahtx** [comment]
> Am I write in thinking that for function wrapped in e.g. `@cached` the cancellation won't propagate to the inner `runInteraction`? (I'm just thinking how this interacts with caching, in terms of if we've done the DB work it would be good to add it to the cache)

Yes, that's right. We shouldn't see a cancellation at this level if we're inside a `@cached` method.

File: tests/storage/test_database.py [hunk 1]
```diff
@@ -22,3 +35,159 @@ def test_native_tuple_comparison(self):
...
+
+        after_callback.assert_not_called()
+        exception_callback.assert_called_once_with(987, 654, extra=321)
+
+    def test_failed_retry(self) -> None:
+        """Test that the exception callback is called for every failed attempt."""
+
+        def _test_txn(txn: LoggingTransaction) -> NoReturn:
+            """Simulate a retryable failure on every attempt."""
+            raise self.db_pool.engine.module.OperationalError()
```
Comment (squahtx) [inline tests/storage/test_database.py:96]: Consider using `Mock(side_effect=self.db_pool.engine.module.OperationalError)`

**Pull request merged**